### DISCUSSION TITLE
Unhandled exception in RNS when using pyinstaller

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#332] [Help & Questions]
I'm using an RNS daemon started at system startup with a systemd unit. It runs under a user account, named x for example. It uses the configuration in /etc/reticulum, which is owned by user x and the error occurs when I run the app as user x. The permissions on the /etc/reticulum directory (and all subfolders) are 755, and files are 644. 

When I run the app normally with the python interpreter all is fine. I use a hard-coded config path string I pass to the RNS.Reticulum constructor of "/etc/reticulum".  I have all of the Interfaces commented out in the config except a TCP Client Interface.

When I run the executable produced by pyinstaller the app produces this error:
```
$ ./PiratePublisher 
Using default config path (~/.config/PriateBoxV2/)
[2023-06-21 23:22:01] [Error] Local shared instance appears to be running, but it could not be connected
[2023-06-21 23:22:01] [Error] The contained exception was: name 'LocalInterface' is not defined
Traceback (most recent call last):
  File "PiratePublisher.py", line 18, in <module>
  File "pboxPublisher.py", line 115, in __init__
  File "reticulum.py", line 53, in __init__
  File "RNS/Reticulum.py", line 255, in __init__
  File "RNS/Reticulum.py", line 391, in __apply_config
NameError: name 'Interface' is not defined
[579249] Failed to execute script 'PiratePublisher' due to unhandled exception!
RNS destroyed
```
I'm using pyInstaller for the first time and have been studying the docs and looking for info online about it but so far I'm unable to get a grip on this issue. I doubt it's an RNS problem, although there are plenty of "ToDo" comments in the Reticulum.py file tthat might impact on this. PyInstaller produces the following warnings related to RNS, but these may not be indicative of actual problems due to conditional imports. 

```
missing module named validate - imported by RNS.vendor.configobj (delayed, conditional)
missing module named compiler - imported by RNS.vendor.configobj (delayed, conditional)
missing module named StringIO - imported by RNS.vendor.six (conditional)
missing module named 'usbserial4a.cdcacmserial4a' - imported by RNS.Interfaces.Android.RNodeInterface (delayed, conditional), RNS.Interfaces.Android.SerialInterface (delayed, conditional), RNS.Interfaces.Android.KISSInterface (delayed, conditional)
missing module named usb4a - imported by RNS.Interfaces.Android.RNodeInterface (delayed, conditional), RNS.Interfaces.Android.SerialInterface (delayed), RNS.Interfaces.Android.KISSInterface (delayed)
missing module named usbserial4a - imported by RNS.Interfaces.Android.RNodeInterface (delayed, conditional), RNS.Interfaces.Android.SerialInterface (delayed, conditional), RNS.Interfaces.Android.KISSInterface (delayed, conditional)
missing module named jnius - imported by RNS.Interfaces.Android.RNodeInterface (delayed)
```

I don't see anything in that list which looks like an issue, since I'm only using the TCP Client Interface. That it complains "Interface" is not defined seems very odd though. 

Posting this hoping others here may have experience using PyInstaller and can offer suggestions on how to resolve this runtime error.

**Comment by cobraPA** [comment]
I haven't used pyinstaller, but this seems partly like a python path issue.  LocalInterface and Interface are modules in the python tree and should always be found, so that does sound like an issue with pyinstaller (or some setup for it you had to do?)

The below is probably not fatal, although if you don't see it launch an instance successfully then it will be:

> Local shared instance appears to be running, but it could not be connected

Are you running an instance of Reticulum already on that machine when this occurs?

**Reply by ThomasFreedman** [reply]
Thanks for your reply. Yes, I run RNS as a systemd service. I suspect the application can't access the /etc/reticulum folder to read the config file. I don't yet understand how to provide the correct options to PyInstaller so it can locate things outside of the virtualized space it creates to run the app in. 

I made an attempt to put the RNS folder in the same place as the app files but that also fails, but in a different way, tho still path related: 
```
$ ./PiratePublisher 
Using default config path
Traceback (most recent call last):
  File "PiratePublisher.py", line 18, in <module>
  File "pboxPublisher.py", line 78, in __init__
  File "logging/__init__.py", line 1999, in basicConfig
  File "logging/__init__.py", line 1142, in __init__
  File "logging/__init__.py", line 1171, in _open
FileNotFoundError: [Errno 2] No such file or directory: '/volumes/Accounts/x/PycharmProjects/test/dist/.config/PirateBoxV2/pbxV2.log'
[634122] Failed to execute script 'PiratePublisher' due to unhandled exception!

```

**Reply by faragher** [reply]
Is the daemon also pyinstaller based or traditional Python?

If you believe it to be a problem with permissions and don't wish to use a user home, you can use /opt/... and then you can be a little more lax with read permissions to test if that's where the problem is.

**Reply by ThomasFreedman** [reply]
On the last trial I didn't use the daemon. I stopped it and moved the reticulum config folder into a subdir of the project. That won't do for a final app release but it will be ok for now to test. That did seem to eliminate the RNS related errors but now it has an error writing to a log file within the project space. The reticulum config folder is under in the .config directory so I am very puzzled by the app log file (not related to reticulum) error.

**Reply by dietervansteenwegen** [reply]
Not sure if this is obvious (sorry if it is), but  do you include the config folder (which you've moved as a subdir of the project)?

Example of how to include files during the "compilation" of pyinstaller (add to the beginning of `main.spec`):

```python
added_files = [
         ('assets/fonts', 'assets/fonts'),
         ('assets/poppler', 'assets/poppler'),
          ('./config.toml', '.'),
          ]
```
The first two are complete directories that get copied in the same subfolders, the third one is a toml file that gets copied to the "root" of the "compiled" project.

I realize that this isn't  your issue, but since you indicate you are trying to troubleshoot by including the files during the creation of the `.exe`, you don't want to have additional things not going the way you expect them to go and add red herrings...
